bitbot/urbandict.go

```diff
@@ -0,0 +1,91 @@
+package bitbot
+
+// partly stolen from https://github.com/dpatrie/urbandictionary
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/whyrusleeping/hellabot"
+)
+
+var UrbanDictionaryTrigger = NamedTrigger{
+	ID:   "urbandict",
+	Help: "Get an urban dictionary issued definition. Usage: !urbd [term]",
+	Condition: func(irc *hbot.Bot, m *hbot.Message) bool {
+		return m.Command == "PRIVMSG" && strings.HasPrefix(m.Trailing, "!urbd")
+	},
+	Action: func(irc *hbot.Bot, m *hbot.Message) bool {
+		resp := urbanDefinition(m.Content)
+		irc.Reply(m, resp)
+		return true
+	},
+}
+
+func urbanDefinition(message string) string {
+	term := strings.SplitAfterN(message, " ", 2)[1] // Strip trigger word
+	res, err := urbanDictQuery(term)
+	if err != nil {
+		log.Println(err)
+		return "The search failed"
+	}
+
+	if len(res.Results) > 0 {
+		return cleanDef(res.Results[0].Definition)
+	}
+	return "No definition for that word"
+}
+
+func cleanDef(def string) string {
+	def = strings.ReplaceAll(def, "[", "")
+	def = strings.ReplaceAll(def, "]", "")
+
+	return def
+}
+
+type SearchResult struct {
+	Type    string `json:"result_type"`
+	Tags    []string
+	Results []Result `json:"list"`
+	Sounds  []string
+}
+
+type Result struct {
+	Author     string
+	Word       string
+	Definition string
+	Example    string
+	Permalink  string
+	Upvote     int `json:"thumbs_up"`
+	Downvote   int `json:"thumbs_down"`
+}
+
+func urbanDictQuery(searchTerm string) (*SearchResult, error) {
+	const baseURL = "http://api.urbandictionary.com/v0/define?term="
+	resp, err := http.Get(baseURL + url.QueryEscape(searchTerm))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
```

Comment:
this reminds me that we should make a generic HTTP client with proper timeouts that can be used for all of our API lookups